### PR TITLE
Refactor: Replace hardcoded messages and errors with messages

### LINKED ISSUE #116: Save error and success messages into an array and replace hardcoded text
## Description
As a developer,
I need hardcoded error messages to be saved in an array,
so that I can avoid bugs related to the usage of hardcoded strings and pay attention to the repeated messages.

On backend save common messages of error and HTTP codes on a JSON
Have an Array of messages instead of using all over the code hardcoded text

Find a way to avoid hardcoded messages, that can be duplicated, with small differences. Make this uniform?!

### Update [Required]
- [ ] Suggestion: Create arrays for both success and error messages joined to replace hardcoded text
- [ ] Suggestion: Have an array for error codes as well to see clearly the meaning of the error codes
- [ ] Fix tests accordingly

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
5 hours

### PR DESCRIPTION
**Pull request by CleverFool77**
### Description
This PR changes the hardcoded errors and messages with constants.

I created a `constants.py` file in which messages have been assigned to variables. I preferred this choice over arrays as in arrays , it is need to keep track of indexes for particular messages or every time search for the message required.This approach can make the system slow as searching over arrays is time-consuming task.
I didn't use dictionary to store the data as it would have again lead to hardcoding of strings when accessing the key in the dictionary.

Having all the constants stored in a constants file makes it easier to maintain and use the constants and also faster as it's only importing the specific variable only. I looked in other orgs to find how they handle this problem and found that they are also using constants for hardcoded strings.

I have made changes in test files as the tests were failing due to mismatch in hardcoded sentences 
```
test file = "This mentorship relation was not accepted"
other files = "The mentorship relation was not accepted"
```

Fixes #116

### Type of Change:
**Delete irrelevant options.**

- Code
- Quality Assurance
- User Interface
- Outreach
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?
Describe the tests you ran to verify your changes. Provide instructions or GIFs so we can reproduce. List any relevant details for your test.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by CleverFool77** [comment]
@isabelcosta I made a PR but sysbot closed it .Please reopen it.

**Comment by isabelcosta** [comment]
@CleverFool77 just an advice, try to give more descriptive names to your branches, `myfork1` or `myfork` aren't good names. For example, you could name this `replace-hardcoded-messages`.

**Comment by CleverFool77** [comment]
@isabelcosta Thanks for telling me from now onward I'll keep that in mind :+1:

**Comment by CleverFool77** [comment]
@isabelcosta Thanks for telling me, from now onwards I'll keep that in mind

**Review by ramitsawhney27 — CHANGES_REQUESTED** [review]
Please fix errors reported in the job failure.
In particular,
test_relations_non_auth (users.test_api_home_statistics.TestHomeStatisticsApi) ... FAIL
Traceback (most recent call last):
  File "/home/travis/build/systers/mentorship-backend/tests/users/test_api_home_statistics.py", line 31, in test_relations_non_auth
    self.assertEqual(expected_response, json.loads(actual_response.data))
AssertionError: {'message': 'The authorization token is missing!'} != {'message': 'The authorization token is missing!.'}
- {'message': 'The authorization token is missing!'}
+ {'message': 'The authorization token is missing!.'}
?                                                 +

Traceback (most recent call last):
  File "/home/travis/build/systers/mentorship-backend/tests/users/test_api_login.py", line 56, in test_user_login_non_verified_user
    self.assertEqual('Please verify your email before login.', response.json.get("message", None))
AssertionError: 'Plea[16 chars]mail before login.' != 'Plea[16 chars]mail before                                             login.'
- Please verify your email before login.
+ Please verify your email before                                             login.

**Comment by CleverFool77** [comment]
@ramitsawhney27 I've updated the PR as requested :+1:

File: app/constants.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+NAME_LENGTH_GREATER_THAN_MAX_LIMIT = "The {field_name} field has to be shorter \
+                                      than {max_limit} characters."
+NAME_LENGTH_LESSER_THAN_MAX_LIMIT = "The {field_name} field has to be longer \
+                                     than {min_limit} characters and \
+                                     shorter than {max_limit} characters."
+USER_INPUTS_INCORRECT_CONFIGURATION_VALUE = "The environment config value has \
+                                            to be within these values: prod, \
+                                            dev, test."
+
+# comment = Mentorship state."
```
Comment (isabelcosta) [inline app/constants.py]: why do you have `comment` keyword in the comment here and in other places of this file?

File: app/api/dao/admin.py [hunk 2]
```diff
@@ -21,21 +21,21 @@ def assign_new_user(assigner_user_id, data):
         new_admin_user_id = data['user_id']
 
         if assigner_user_id is new_admin_user_id:
-            return {"message": "You cannot assign yourself as an Admin."}, 403
+            return {"message": constants.USER_CANNOT_ASSIGN_HIMSELF_AS_ADMIN}, 403
```
Comment (isabelcosta) [inline app/api/dao/admin.py]:
```suggestion
            return {"message": constants.USER_CANNOT_ASSIGN_SELF_AS_ADMIN}, 403
```
I'm not sure if this is a good correction, but avoid using pronouns in your code, pronouns that assume the gender of the user. I haven't check this on other places, but be aware of this. Do a search for pronouns in the rest of your commit.

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Great job @CleverFool77 !
I left some comments for you to address.

**Comment by isabelcosta** [comment]
@ramitsawhney27 it makes sense to change the hardcoded messages and use these constants on the tests module as well?
cc @CleverFool77

**Comment by ramitsawhney27** [comment]
Yes, all hardcoded instances should be replaced now as we move to a modular approach.

**Review by ramitsawhney27 — CHANGES_REQUESTED** [review]
Why not replace the entire json object in the constants file? It’s extra replication. Rather have the dicts themselves in the constants module. Also, please use a more informative name than just constants.

**Comment by isabelcosta** [comment]
> Why not replace the entire json object in the constants file? It’s extra replication. Rather have the dicts themselves in the constants module.

@ramitsawhney27 should we then have a function that receives the constant message and returns the json object? Like:
```
def createResponse(message):
     return {"message": message}

createResponse(USER_DOES_NOT_EXIST)
```

In this way, "message" keyword is not repeated and if in the future we want to add something to the response, we can edit here in the function?

cc @CleverFool77

File: app/api/dao/admin.py (hunk 2 above)
Comment (CleverFool77) [reply app/api/dao/admin.py]: @isabelcosta I'll remove all the pronouns :+1:

**Comment by CleverFool77** [comment]
@ramitsawhney27 @isabelcosta Not always the format of json "message" and then the message is used. In some places, format like this is also used. 
```
self.assertEqual(constants.USER_HAS_NOT_VERIFIED_EMAIL_BEFORE_LOGIN,
```

File: app/constants.py (hunk 1 above)
Comment (CleverFool77) [reply app/constants.py]: I'll resolve this @isabelcosta :+1:

**Comment by CleverFool77** [comment]
@isabelcosta @ramitsawhney27 I've updated the PR as requested :+1:

**Comment by ramitsawhney27** [comment]
Overhead for calling a function is way more than just directly using a constant literal. For something so repetitive, I'd rather use literals in a variables file.

**Comment by isabelcosta** [comment]
@ramitsawhney27 I totally trust you on this, I just need to know how that translates to code. @CleverFool77 did you understand what @ramitsawhney27 is asking? Can you do that?

Do you mean doing this like:
```
USER_DOES_NOT_EXIST = {"message": "This user does not exist."}
```

**Comment by CleverFool77** [comment]
I've used constants literals for string as @ramitsawhney27 suggested. I don't think it's good idea to assign dictionary to each constant since in the test string output is matched rather than dictionary object. @isabelcosta 
Eg :
```python
USER_DOES_NOT_EXIST = {"message": "This user does not exist."}
```
This can't be tested with 
```python
self.assertEqual(constants.USER_HAS_NOT_VERIFIED_EMAIL_BEFORE_LOGIN,response.json.get("message", None))
```
In my opinion, Comparing two string is much better than comparing two dictionary objects

**Comment by ramitsawhney27** [comment]
Testing is pre-production, not running at the time of production. If the tests take longer, it's not a problem.

**Comment by CleverFool77** [comment]
@ramitsawhney27 I'll update the PR soon as requested :+1:

**Comment by CleverFool77** [comment]
@ramitsawhney27 @isabelcosta The PR is updated as requested :+1: 
* json object is assigned to constants to avoid extra replication. 
* Changed tests as per above discussion.

**Review by ramitsawhney27 — DISMISSED** [review]
Good work, thank you for contributing.

**Comment by isabelcosta** [comment]
@ramitsawhney27 I remember you saying this:

> Also, please use a more informative name than just constants.

Would `messages` be a better name?

**Comment by CleverFool77** [comment]
@isabelcosta should I go for name `messages` instead of `constants`.

**Comment by CleverFool77** [comment]
@isabelcosta @ramitsawhney27 The PR is updated as requested :+1: 
* A more informative name `messages` is used instead of constants.

**Review by isabelcosta — DISMISSED** [review]
Great work on this refactor @CleverFool77 🎉

**Comment by isabelcosta** [comment]
@CleverFool77 I merged two PRs that generated merge conflicts in your PR, can you please solve them so that I can review and merge your PR again? And squash your commits in the end.

**Comment by CleverFool77** [comment]
@isabelcosta I've removed the merge conflicts. This PR will create merge conflicts for some other PRs due to refactoring in hardcoded strings format.

**Review by ramitsawhney27 — DISMISSED** [review]
Great work, thanks for contributing!

**Review by isabelcosta — DISMISSED** [review]
Thank you for fixing the merge conflicts @CleverFool77

**Comment by isabelcosta** [comment]
Uh sorry I just noticed there are some messages mismatched @CleverFool77 I will point some, and let you look at other messages.

File: app/api/dao/admin.py [hunk 3]
```diff
@@ -21,21 +21,21 @@ def assign_new_user(assigner_user_id, data):
...
 
             if new_admin_user.is_admin:
-                return {"message": "User is already an Admin."}, 400
+                return messages.USER_IS_ALREADY_AN_ADMIN, 400
 
             new_admin_user.is_admin = True
             new_admin_user.save_to_db()
 
-            return {"message": "User is now an Admin."}, 200
+            return messages.USER_IS_ALREADY_AN_ADMIN, 200
```
Comment (isabelcosta) [inline app/api/dao/admin.py]: This is wrong. This should point to the successful message of "User is now an Admin", and you're using the message "User is already an Admin"

File: app/messages.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                                                " successfully."}
+MENTORSHIP_RELATION_WAS_REJECTED_SUCCESSFULLY = {"message": "Mentorship"
+                                                 " relation was"
+                                                 " rejected successfully."}
+MENTORSHIP_RELATION_WAS_CANCELLED_SUCCESSFULLY = {"message": "Mentorship"
+                                                  " relation was "
+                                                  "cancelled successfully"}
+TASK_WAS_CREATED_SUCCESSFULLY = {"message": "Task was created successfully."}
+TASK_WAS_DELETED_SUCCESSFULLY = {"message": "Task was deleted successfully."}
+TASK_WAS_ALREADY_ACHIEVED_SUCCESSFULLY = {"message": "Task was achieved"
```
Comment (isabelcosta) [inline app/messages.py]:
```suggestion
TASK_WAS_ACHIEVED_SUCCESSFULLY = {"message": "Task was achieved"
```

File: app/messages.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+MENTOR_TIME_GREATER_THAN_MAX_TIME = {"message": "Mentorship relation maximum"
+                                     " duration is 6 months."}
+MENTOR_TIME_LESS_THAN_MIN_TIME = {"message": "Mentorship relation minimum"
+                                  " duration is 4 week."}
+CANT_ACCEPT_MENTOR_REQ_SENT_BY_USER = {"message": "You cannot accept a"
+                                       " mentorship request sent by yourself."}
+CANT_ACCEPT_UNINVOLVED_MENTOR_RELATION = {"message": "You cannot accept a"
+                                          " mentorship relation where you are"
+                                          " not involved."}
+USER_CANT_REJECT_REQUEST_FOR_MENTOR = {"message": "You cannot reject a"
```
Comment (isabelcosta) [inline app/messages.py]:
```suggestion
USER_CANT_REJECT_REQUEST_SENT_BY_USER = {"message": "You cannot reject a"
```

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@CleverFool77 I left some comments for you to address. You're almost there ;)
Please take a closer look at the messages you're assigning, since you mismatched at least one message of error with success.

File: app/api/dao/admin.py (hunk 3 above)
Comment (CleverFool77) [reply app/api/dao/admin.py]: sure :+1:

File: app/messages.py (hunk 4 above)
Comment (CleverFool77) [reply app/messages.py]: I'll resolve this :+1:

**Comment by CleverFool77** [comment]
@isabelcosta I've updated the PR :smiley:

File: app/messages.py (hunk 5 above)
Comment (CleverFool77) [reply app/messages.py]: Done :+1:

**Review by isabelcosta — DISMISSED** [review]

File: tests/admin/test_dao.py [hunk 6]
```diff
@@ -5,6 +5,7 @@
 from app.database.models.user import UserModel
 from app.api.dao.admin import AdminDAO
 from app.database.sqlalchemy_extension import db
+from app import messages
```
Comment (iulianav) [inline tests/admin/test_dao.py]: I think this import should be the first one? Anyway, I see the imports are not in the correct PEP8 compliant alphabetical order.

File: tests/mentorship_relation/test_api_accept_request.py [hunk 7]
```diff
@@ -8,7 +8,7 @@
 from app.utils.enum_utils import MentorshipRelationState
 from tests.mentorship_relation.relation_base_setup import MentorshipRelationBaseTestCase
 from tests.test_utils import get_test_request_header
-
+from app import messages
```
Comment (iulianav) [inline tests/mentorship_relation/test_api_accept_request.py]: Same problem, wrong position for this import.

File: tests/mentorship_relation/test_api_cancel_relation.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
 from app.utils.enum_utils import MentorshipRelationState
 from tests.mentorship_relation.relation_base_setup import MentorshipRelationBaseTestCase
 from tests.test_utils import get_test_request_header
+from app import messages
```
Comment (iulianav) [inline tests/mentorship_relation/test_api_cancel_relation.py]: Wrong position for import.

File: tests/mentorship_relation/test_api_reject_request.py [hunk 9]
```diff
@@ -8,6 +8,7 @@
 from app.utils.enum_utils import MentorshipRelationState
 from tests.mentorship_relation.relation_base_setup import MentorshipRelationBaseTestCase
 from tests.test_utils import get_test_request_header
+from app import messages
```
Comment (iulianav) [inline tests/mentorship_relation/test_api_reject_request.py]: Wrong position for import.

File: tests/mentorship_relation/test_dao_accept_request.py [hunk 10]
```diff
@@ -6,6 +6,7 @@
 from app.utils.enum_utils import MentorshipRelationState
 from tests.mentorship_relation.relation_base_setup import MentorshipRelationBaseTestCase
 from app.database.sqlalchemy_extension import db
+from app import messages
```
Comment (iulianav) [inline tests/mentorship_relation/test_dao_accept_request.py]: Wrong position for import.

File: tests/mentorship_relation/test_dao_cancel_relation.py [hunk 11]
```diff
@@ -6,6 +6,7 @@
 from app.database.models.mentorship_relation import MentorshipRelationModel
 from app.database.sqlalchemy_extension import db
 from tests.mentorship_relation.relation_base_setup import MentorshipRelationBaseTestCase
+from app import messages
```
Comment (iulianav) [inline tests/mentorship_relation/test_dao_cancel_relation.py]: Wrong position for import

File: tests/mentorship_relation/test_dao_creation.py [hunk 12]
```diff
@@ -7,6 +7,7 @@
 from app.utils.enum_utils import MentorshipRelationState
 from tests.mentorship_relation.relation_base_setup import MentorshipRelationBaseTestCase
 from app.database.sqlalchemy_extension import db
+from app import messages
```
Comment (iulianav) [inline tests/mentorship_relation/test_dao_creation.py]: Wrong ordering for import.

**Comment by CleverFool77** [comment]
@iulianav I was told about import order in my other PR too. Currently I have my exam so I haven't updated them but yeah I'll update them soon in few days. Thanks for reviewing :+1: and helping me to make it a Quality PR.

**Comment by iulianav** [comment]
@CleverFool77 don't worry and good luck with your exams! From experience reviews are great learning experiences for both the reviewer and the reviewed!

**Comment by isabelcosta** [comment]
@iulianav thank you for your review here 🎉 These things I tend not to notice since I'm not familiar with PEP8 standards.

@CleverFool77 do this when you're able to :)

**Comment by iulianav** [comment]
@isabelcosta happy to help anytime!

**Comment by CleverFool77** [comment]
@iulianav @isabelcosta I've updated the PR.

File: app/messages.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+                         " revoke other admin user."}
+USER_IS_NOW_AN_ADMIN = {"message": "User is now an Admin."}
+
+# Mentor availability
+MENTOR_NOT_AVAILABLE_TO_MENTOR = {"message": "Mentor user is not available to"
+                                  " mentor."}
+MENTOR_IN_RELATION = {"message": "Mentor user is already in a relationship."}
+
+# Mentee availability
+MENTEE_NOT_AVAIL_TOBE_MENTORED = {"message": "Mentee user is not available"
```
Comment (isabelcosta) [inline app/messages.py]:
```suggestion
MENTEE_NOT_AVAIL_TO_BE_MENTORED = {"message": "Mentee user is not available"
```

**Review by isabelcosta — DISMISSED** [review]
@CleverFool77 I left a comment that I would appreciate if you would address it, but from the rest I think the rest of the code is good :)

**Comment by CleverFool77** [comment]
@isabelcosta I've updated the PR.

**Review by isabelcosta — APPROVED** [review]

**Pull request merged**